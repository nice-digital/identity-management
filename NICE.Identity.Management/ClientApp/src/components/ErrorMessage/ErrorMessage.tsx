import React from "react";

import { Alert } from "@nice-digital/nds-alert";

type ErrorMessageProps = {
	error: Error;
	children?: React.ReactNode;
};

export const ErrorMessage = (props: ErrorMessageProps): React.ReactElement => (
	<Alert type="error">
		{props.children ? (
			props.children
		) : (
			<>
				<h2>Whoops</h2>
				<p>Sorry, there&apos;s been an error!</p>
			</>
		)}

		{process.env.NODE_ENV === "development" && (
			<dl>
				<dt>Message:</dt>
				<dd>{props.error.message}</dd>
				<dt>Stack:</dt>
				<dd>{props.error.stack}</dd>
			</dl>
		)}
	</Alert>
);
