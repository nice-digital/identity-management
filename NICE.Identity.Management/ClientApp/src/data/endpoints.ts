const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
