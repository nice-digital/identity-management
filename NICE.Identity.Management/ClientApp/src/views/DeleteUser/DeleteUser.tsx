import React, { Component } from "react";
import { RouteComponentProps, Link } from "react-router-dom";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Button } from "@nice-digital/nds-button";
import { PageHeader } from "@nice-digital/nds-page-header";

import { DeleteUserConfirmation } from "../DeleteUserConfirmation/DeleteUserConfirmation";
import { UserType } from "../../models/types";
import { Endpoints } from "../../data/endpoints";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

import styles from "./DeleteUser.module.scss";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";

type TParams = { id: string };

type DeleteUserProps = Record<string, unknown> & RouteComponentProps<TParams>;

type DeleteUserState = {
	user: UserType;
	hasBeenDeleted: boolean;
	error?: Error;
	isLoading: boolean;
	isDeleteButtonLoading: boolean;
};

export class DeleteUser extends Component<DeleteUserProps, DeleteUserState> {
	constructor(props: DeleteUserProps) {
		super(props);

		this.state = {
			user: {} as UserType,
			hasBeenDeleted: false,
			isLoading: true,
			isDeleteButtonLoading: false,
		};
	}

	handleDeleteClick = async (): Promise<void> => {
		this.setState({ isDeleteButtonLoading: true });

		const fetchOptions = {
			method: "DELETE",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({ userId: this.props.match.params.id }),
		};

		const deletedUser = await fetchData(
			Endpoints.user(this.props.match.params.id),
			fetchOptions,
		);

		if (isDataError(deletedUser)) {
			this.setState({ error: deletedUser });
		} else {
			this.setState({ hasBeenDeleted: true });
		}

		this.setState({ isDeleteButtonLoading: false });
	};

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const user = await fetchData(Endpoints.user(this.props.match.params.id));

		if (isDataError(user)) {
			this.setState({ error: user });
		}

		this.setState({ user, isLoading: false });
		document.title = `NICE Accounts - Delete ${user.firstName} ${user.lastName}`;
	}

	render(): JSX.Element {
		const { isLoading, isDeleteButtonLoading, error, hasBeenDeleted, user } =
			this.state;
		const { id } = this.props.match.params;

		let lastBreadcrumb = `${user.firstName} ${user.lastName}`;

		if (isLoading) {
			lastBreadcrumb = "Loading user details";
		}

		return (
			<>
				{hasBeenDeleted ? (
					<DeleteUserConfirmation
						fullName={`${user.firstName} ${user.lastName}`}
					/>
				) : (
					<>
						<Breadcrumbs>
							<Breadcrumb to="/users" elementType={Link}>
								Users
							</Breadcrumb>
							{error ? (
								<Breadcrumb>Error</Breadcrumb>
							) : (
								<Breadcrumb to={`/users/${id}`} elementType={Link}>
									{lastBreadcrumb}
								</Breadcrumb>
							)}
							<Breadcrumb>Delete user</Breadcrumb>
						</Breadcrumbs>

						{!error ? (
							<>
								{isLoading ? (
									<>
										<PageHeader preheading="Confirm" heading="Delete user" />
										<p>Loading...</p>
									</>
								) : (
									<>
										<PageHeader
											preheading="Are you sure you want to delete user?"
											heading={`${user.firstName} ${user.lastName}`}
											cta={
												<>
													<Button
														data-qa-sel="confirm-delete-user"
														variant="cta"
														onClick={this.handleDeleteClick}
														disabled={isDeleteButtonLoading}
													>
														{isDeleteButtonLoading ? "Loading..." : "Confirm"}
													</Button>
													<Button
														to={`/users/${id}`}
														variant="secondary"
														elementType={Link}
														disabled={isDeleteButtonLoading}
													>
														Cancel
													</Button>
												</>
											}
										/>
										<hr className="mv--b" />
										<Grid className="pv--d">
											<GridItem cols={3}>
												<span className={styles.detailsLabel}>
													Email address
												</span>
											</GridItem>
											<GridItem cols={9}>
												<span>{user.emailAddress}</span>
											</GridItem>
										</Grid>
									</>
								)}
							</>
						) : (
							<>
								<PageHeader heading="Error" />
								<ErrorMessage error={error}></ErrorMessage>
							</>
						)}
					</>
				)}
			</>
		);
	}
}
