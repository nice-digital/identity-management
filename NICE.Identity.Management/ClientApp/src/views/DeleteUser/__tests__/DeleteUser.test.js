import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../utils/nextTick";
import singleUser from "./singleUser.json";
import { DeleteUser } from "../DeleteUser";

import { Endpoints } from "../../../data/endpoints";

import { ErrorMessage } from "../../../components/ErrorMessage/ErrorMessage";

describe("DeleteUser", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});


	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = shallow(<DeleteUser match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetchData during componentDidMount", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = mount(<MemoryRouter><DeleteUser match={match} /></MemoryRouter>);
		const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
		wrapper.instance().componentDidMount();
		wrapper.update();
		expect(spy).toHaveBeenCalled();
		expect(fetch.mock.calls.length).toEqual(1);
		expect(fetch.mock.calls[0][0]).toEqual(Endpoints.user(match.params.id));
		spy.mockClear();
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = shallow(<DeleteUser match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetchData function returns 401 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetchData function returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetchData delete fails", async () => {
		console.error = jest.fn();
		const error = new Error("Not allowed");
		fetch.mockResponseOnce(JSON.stringify({}));
		fetch.mockRejectOnce(error);
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("button").simulate("click");
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when fetchData delete returns non-200 error", async () => {
		console.error = jest.fn();
		const serverErrorMessage = "Not authorized";
		fetch.mockResponseOnce(JSON.stringify({}));
		fetch.mockResponseOnce({ message: serverErrorMessage }, { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("button").simulate("click");
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should disable delete button when clicked", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify({}));
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("button").simulate("click");
		expect(wrapper.find("button").props().disabled).toEqual(true);
		expect(wrapper.find("button").text()).toEqual("Loading...");
	});

	it("should display confirmation message once fetchData delete is successfully complete", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify({}));
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("button").simulate("click");
		await nextTick();
		wrapper.update();
		expect(wrapper.find("h1").text()).toEqual("User deleted");
	});
});
