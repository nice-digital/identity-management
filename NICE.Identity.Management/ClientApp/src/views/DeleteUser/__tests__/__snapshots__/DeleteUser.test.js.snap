// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteUser should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/users/1"
    >
      Kristin Patrick
    </Breadcrumb>
    <Breadcrumb>
      Delete user
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    cta={
      <React.Fragment>
        <Button
          data-qa-sel="confirm-delete-user"
          disabled={false}
          onClick={[Function]}
          variant="cta"
        >
          Confirm
        </Button>
        <Button
          disabled={false}
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/users/1"
          variant="secondary"
        >
          Cancel
        </Button>
      </React.Fragment>
    }
    heading="Kristin Patrick"
    preheading="Are you sure you want to delete user?"
  />
  <hr
    className="mv--b"
  />
  <Grid
    className="pv--d"
    elementType="div"
    equalHeight={false}
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      cols={3}
    >
      <span
        className="detailsLabel"
      >
        Email address
      </span>
    </GridItem>
    <GridItem
      cols={9}
    >
      <span>
        kristin.patrick@example.com
      </span>
    </GridItem>
  </Grid>
</Fragment>
`;

exports[`DeleteUser should show error message when fetchData function returns 401 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/users"
            onClick={[Function]}
          >
            Users
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Error
          </span>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Delete user
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/users\\",
        \\"name\\": \\"Users\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Error\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 3,
      \\"item\\": {
        \\"name\\": \\"Delete user\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Error
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;

exports[`DeleteUser should show error message when fetchData function returns 500 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/users"
            onClick={[Function]}
          >
            Users
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Error
          </span>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Delete user
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/users\\",
        \\"name\\": \\"Users\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Error\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 3,
      \\"item\\": {
        \\"name\\": \\"Delete user\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Error
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;
